File: neet_code_150_cpp/bit_manipulation/single_number.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
    public:
        // 縛り無視ならこう実装する
        // AC
        int singleNumberMap(std::vector<int>& nums) {
            std::unordered_map<int, int> map;

            for(auto &v: nums) {
                ++map[v];
            }

            int result = 0;
            for(auto &v: map) {
                if(v.second == 1) {
                    result = v.first;
                    break;
                }
            }

            return result;
        }
};

int main(void) {
    auto case_1 = std::vector{2, 2, 1};
    auto case_2 = std::vector{4, 1, 2, 2, 1};
    auto case_3 = std::vector{1};

    Solution s_1;
    std::cout << s_1.singleNumberMap(case_1) << std::endl;
    std::cout << s_1.singleNumberMap(case_2) << std::endl;
    std::cout << s_1.singleNumberMap(case_3) << std::endl;
}

File: neet_code_150_go/bit_manipulation/single_number.go
package main

import "fmt"

// AC
func singleNumber(nums []int) int {
	result := 0
	for _, n := range nums {
		result ^= n
	}

	return result
}

func main() {
	case_1 := []int{2, 2, 1}
	case_2 := []int{4, 1, 2, 1, 2}
	case_3 := []int{1}

	fmt.Println("case_1: ", singleNumber(case_1))
	fmt.Println("case_2: ", singleNumber(case_2))
	fmt.Println("case_3: ", singleNumber(case_3))
}
